pennylane_lightning/experimental/_measure.py

```diff
@@ -0,0 +1,136 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Internal code relevant for performing measurements on a state.
+"""
+
+import numpy as np
+from typing import Callable, List
+
+from pennylane.measurements import StateMeasurement, MeasurementProcess, ExpectationMP
+from pennylane.typing import TensorLike
+from pennylane.wires import Wires
+
+from ._apply_operations import apply_operations
+
+from ..lightning_qubit_ops import (
+    MeasuresC64,
+    StateVectorC64,
+    MeasuresC128,
+    StateVectorC128,
+    Kokkos_info,
+)
+
+from ._serialize import _serialize_ob
+
+
+def state_diagonalizing_gates(measurementprocess: StateMeasurement, state: np.array) -> TensorLike:
+    """Apply a measurement to state when the measurement process has an observable with diagonalizing gates.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (np.array): unravelled state (1D) to apply the measurement to
+
+    Returns:
+        TensorLike: the result of the measurement
+    """
+    state = apply_operations(measurementprocess.diagonalizing_gates(), np.copy(state))
+
+    total_wires = state.size >> 1
+    wires = Wires(range(total_wires))
+    return measurementprocess.process_state(state, wires)
+
+
+def expval(measurementprocess: MeasurementProcess, state: np.array):
+    """Expectation value of the supplied observable contained in the MeasurementProcess.
+
+    Args:
+        measurementprocess (StateMeasurement): measurement to apply to the state
+        state (np.array): unravelled state (1D) to apply the measurement to
+
+    Returns:
+        Expectation value of the observable
+    """
+    if state.dtype == np.complex64:
+        state_vector = StateVectorC64(state)
+        M = MeasuresC64(state_vector)
+    else:
+        state_vector = StateVectorC128(state)
+        M = MeasuresC128(state_vector)
+
+    if measurementprocess.obs.name == "SparseHamiltonian":
+        if Kokkos_info()["USE_KOKKOS"] == True:
+            # ensuring CSR sparse representation.
+            total_wires = state.size >> 1
```

Comment:
Same herre as above, this just divides the number of coeffs by 2. Should this be log2 instead for the wire count?

Reply:
:astonished: same here